src/main/java/sg/dex/starfish/keeper/DirectPurchaseAdapter.java

```diff
@@ -0,0 +1,111 @@
+package sg.dex.starfish.keeper;
+
+import sg.dex.starfish.impl.squid.SquidService;
+import com.oceanprotocol.common.web3.KeeperService;
+
+import com.oceanprotocol.squid.exceptions.TokenApproveException;
+import java.io.IOException;
+import org.web3j.crypto.CipherException;
+
+import org.web3j.utils.Numeric;
+import java.util.ArrayList;
+import java.math.BigInteger;
+import java.util.Properties;
+import java.io.InputStream;
+import com.oceanprotocol.keeper.contracts.OceanToken;
+import org.web3j.protocol.core.methods.response.TransactionReceipt;
+
+import org.web3j.protocol.core.methods.request.EthFilter;
+import io.reactivex.Flowable;
+import org.web3j.abi.EventEncoder;
+import org.web3j.protocol.core.DefaultBlockParameter;
+import org.web3j.protocol.core.DefaultBlockParameterName;
+
+public final class DirectPurchaseAdapter {
+    private DirectPurchase directPurchase;
+    private TokenManager tokenManager;
+
+    public DirectPurchaseAdapter() throws IOException, CipherException {
+        // getting properties
+        Properties properties = getProperties();
+        String directPurchaseAddress = (String)properties.getOrDefault("contract.DirectPurchase.address", "");
+        String oceanTokenAddress = (String)properties.getOrDefault("contract.OceanToken.address", "");
+        // getting keeper
+        KeeperService keeper = SquidService.getKeeperService(properties);
+        // loading contract instances
+        directPurchase = DirectPurchase.load(directPurchaseAddress, keeper.getWeb3(), keeper.getTxManager(), keeper.getContractGasProvider());
+        OceanToken oceanToken = OceanToken.load(oceanTokenAddress, keeper.getWeb3(), keeper.getTxManager(), keeper.getContractGasProvider());
+        // initializing token manager
+        tokenManager = TokenManager.getInstance(keeper);
+        tokenManager.setTokenContract(oceanToken);
+    }
+
+    public TransactionReceipt sendTokenAndLog(String to, BigInteger amount, String reference1, String reference2) {
+        boolean tokenApproved = false;
+        TransactionReceipt receipt = null;
+        try {
+            tokenApproved = tokenApprove(directPurchase.getContractAddress(), amount.toString());
+        } catch (TokenApproveException e) {
+            e.printStackTrace();
+            return receipt;
+        }
+
+        if(tokenApproved) {
+            try {
+                byte[] ref1 = Numeric.hexStringToByteArray(Numeric.toHexStringWithPrefixZeroPadded(Numeric.toBigInt(reference1), 64));
```

Comment:
Prefer to use our Hex functions in `sg.dex.starfish.util` for consistency. If you need new functions there please add them?

I don't really like the conversion to BigInts and then back to Strings. Why do this? If you are worried about users passing in a bad input, I would prefer to throw an `IllegalArgumentException`.